### PR TITLE
Add SelectNode and Remove method to JsonDynamicObject class

### LINKED ISSUE #15505: Inconsistent JObject.SelectNode() behavior and feature loss
### Describe the bug

I have a few concerns with the new `JObject.SelectNode()` extension method:

If you try to use it on a content part's `Content` property, it won't work unless you re-serialize the JSON object first, e.g.:
```csharp
var content = (JsonObject)contentElement.Content;
content.SelectNode(path); // This is null.
JsonSerializer.SerializeToNode(content).SelectNode(path); // This is not null.
```

Also you can't chain `SelectNode` calls, the second call will still search from the same position, e.g.:
image

I think when you query a `JsonObject`, it tries to query from its document root instead of the current element. This is unexpected.

Finally, I'm not sure if anyone is affected, but after the migration to STJ we have lost full compliant JSONPath query support. The documentation of `JObject.SelectNode()` extension method [claims that it searches using "a JSON path"](https://github.com/OrchardCMS/OrchardCore/blob/50583b0644fe90f0a7253c78885504c910c4175a/src/OrchardCore/OrchardCore.Abstractions/Json/Nodes/JObject.cs#L89), however this is misleading because it's compliant with [the JSONPath specification](https://www.ietf.org/archive/id/draft-goessner-dispatch-jsonpath-00.html). An example of a query that should work but doesn't (using the same JSON as above):

```csharp
var data = JsonNode.Parse("{\"SelectedContentItem\":{\"ContentItemIds\":[\"48yrgcwkhv7hkwarvtzpw3ccms\"]},\"Name\":\"My Content\"}");
data.SelectNode("SelectedContentItem.ContentItemIds"); // works, result: [ "48yrgcwkhv7hkwarvtzpw3ccms" ]
data.SelectNode("$.SelectedContentItem.ContentItemIds"); // works, result: [ "48yrgcwkhv7hkwarvtzpw3ccms" ]
data.SelectNode("$..ContentItemIds"); // doesn't work, result is null but it should be the same as above
```
You can verify that the last one should work using the [JSONPath Online Evaluator](https://jsonpath.com/):
image Also it works with Newtonsoft:

image

### Expected behavior

- `SelectNode` should query from the current element.
- Ideally it should keep supporting full JSONPath.

### Suggestion

Beyond fixing `JObject.SelectNode` to search from the current node, the documentation should be updated:
- The documentation comment I linked above should include a `<remarks>` section explaining what is and what isn't supported by the current implementation.
- The 1.9 release notes should mention this breaking change in [the "Drop Newtonsoft.Json Support" section](https://github.com/OrchardCMS/OrchardCore/blob/50583b0644fe90f0a7253c78885504c910c4175a/src/docs/releases/1.9.0.md#drop-newtonsoftjson-support).

Consider if there is a JSONPath library we can use instead of a homebrew implementation. I haven't used it yet, but [JsonPath.Net](https://www.nuget.org/packages/JsonPath.Net/0.1.3) of the [json-everything](https://github.com/gregsdennis/json-everything) project might be suitable? It's for STJ, MIT licensed and it seems to be actively maintained.

### PR DESCRIPTION
**Pull request by hyzx86**
try to fix https://github.com/OrchardCMS/OrchardCore/issues/15505

### CODE REVIEW

**Comment by hyzx86** [comment]
and #15497

**Comment by hyzx86** [comment]
Hi @hishamco @MikeAlhayek  ， 

Since the Content property could be of type JsonDynamicObject or JsonObject, do we need to migrate all functions that can be called from JsonObject here?

https://github.com/OrchardCMS/OrchardCore/blob/b83e2dc85401fc567eb4b12151e2fadcb7f27360/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/ContentElement.cs#L26-L31

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:90]:
```suggestion
    public bool Remove(string key) => _jsonObject.Remove(key);

    public JsonNode? SelectNode(string? path) => _jsonObject.SelectNode(path);
```

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:96]:
```suggestion
    public JsonNode? SelectToken(string? path) => _jsonObject.SelectNode(path);
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by hyzx86** [comment]
Hi @sarahelsaig  , Can you help verify this PR?

**Comment by hyzx86** [comment]
This PR can't to fix  #15497 , because there was a logical error in the last change

File: test/OrchardCore.Tests/Data/ContentItemTests.cs
Comment (sarahelsaig) [inline test/OrchardCore.Tests/Data/ContentItemTests.cs:125]: To clarify the problems in #15505, these should pass if added at the end of this test method:
```csharp
            Assert.NotNull(((JsonObject)contentItem.As<MyPart>().Content).SelectNode("Text")); // fails

            Assert.NotNull((JsonNode)contentItem.Content.SelectNode("MyPart.Text")); // passes
            Assert.NotNull(JObject.Parse(node.ToJsonString()).SelectNode("Text")); // passes
            Assert.NotNull(node.SelectNode("Text")); // fails
```

The first line is what I mentioned at the start of the issue ("If you try to use it on a content part's Content property, it won't work unless you re-serialize the JSON object first").

The other three lines show that chaining "should" work but doesn't. I think the PR only fixes #15505 if these all pass.

File: src/OrchardCore.Modules/OrchardCore.Menu/Controllers/AdminController.cs
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Menu/Controllers/AdminController.cs:282]: Do we have to change the code if `Remove` is added in this class?

**Comment by sebastienros** [comment]
Fixes https://github.com/OrchardCMS/OrchardCore/issues/15497 too?

File: src/OrchardCore.Modules/OrchardCore.Menu/Controllers/AdminController.cs
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Menu/Controllers/AdminController.cs:282]: The problem here is not because of the Remove function, because `menu.Content` is a document object, so we can not directly remove a `MenuItem` from the document object based on the `menuId`
The code on the main branch here is wrong

**Comment by hyzx86** [comment]
> Fixes #15497 too? 

Yes , I have merged @hishamco 's PR

File: test/OrchardCore.Tests/Data/ContentItemTests.cs
Comment (sarahelsaig) [reply test/OrchardCore.Tests/Data/ContentItemTests.cs:125]: > you can't use SelectNode to find a path from a child object.

Yes, that's the bug. Like I've showed in the issue, it works with Newtonsoft's SelectToken so it should continue working with STJ as well.

**Comment by hishamco** [comment]
Please resolve the conflict

**Comment by sebastienros** [comment]
I'll trust @sarahelsaig on merging this one or not.

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs
Comment (sarahelsaig) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:83]: I like the idea of having the remove method, but it should take `_dictionary` into consideration. Right now if you access a key with `GetValue` and then remove it, then accessing it again with `GetValue` will still return the old property. In other words a test such as this should pass:

```csharp
        [Fact]
        public void RemovingPropertyShouldWork()
        {
            var contentItem = new ContentItem();
            contentItem.GetOrCreate<MyPart>();
            contentItem.Alter<MyPart>(x => x.Text = "test");

            JsonDynamicObject content = contentItem.Content;
            Assert.Null(content.GetValue("not real property")); // Properties that don't exist return null.
            Assert.NotNull(content.GetValue(nameof(MyPart))); // Right now this property exists.

            content.Remove(nameof(MyPart));
            Assert.Null(content.GetValue(nameof(MyPart)));
        }
```

This is what you see if you break at the last line of the above test:
image


The same applies to `JsonDynamicArray` as well.

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
@hyzx86 please resolve the conflict so we can merge this after @sarahelsaig approval

**Comment by hyzx86** [comment]
> @hyzx86 please resolve the conflict so we can merge this after @sarahelsaig approval

Resolved

**Review by sarahelsaig — CHANGES_REQUESTED** [review]
Please only request a review when the checks are successful. You have these errors:
image

Also you didn't address this: https://github.com/OrchardCMS/OrchardCore/pull/15509/files/2e877408abb79e7f6613ce5a091cf918bb99d758#r1534515122

**Comment by hyzx86** [comment]
@sarahelsaig updated .

**Comment by Piedone** [comment]
Given that https://github.com/OrchardCMS/OrchardCore/pull/15524 was merged, is this still needed?

**Comment by hyzx86** [comment]
> Given that #15524 was merged, is this still needed? 

Yes, the PR just modify ` JsonObject. SelectNode ` function implementation, here will repair some direct call ` ContentItem. Content. Remove ` and ` ContentItem. Content. SelectNode `

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs
Comment (sarahelsaig) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:89]: No need for double newlines.
```suggestion

```

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicArray.cs
Comment (sarahelsaig) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicArray.cs:44]: [As I mentioned](https://github.com/OrchardCMS/OrchardCore/pull/15509#:~:text=The%20same%20applies%20to%20JsonDynamicArray%20as%20well.) Please update `_dictionary` in `JsonDynamicArray.Remove` as well.

**Review by sarahelsaig — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicArray.cs
Comment (hyzx86) [reply src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicArray.cs:44]: Updated

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:88]:
```suggestion

    public JsonNode? SelectNode(string path) => _jsonObject.SelectNode(path);
```

File: src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:86]:
```suggestion

        return _jsonObject.Remove(key);
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: test/OrchardCore.Tests/Data/ContentItemTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Data/ContentItemTests.cs:128]: We could use `var`

**Review by hishamco — APPROVED** [review]

File: test/OrchardCore.Tests/Data/ContentItemTests.cs
Comment (hyzx86) [reply test/OrchardCore.Tests/Data/ContentItemTests.cs:128]: The `contentItem.Content` is a dynamic type , specifying the type explicitly here allows for efficient code prompting and compile-time validation

Comment (hishamco) [reply test/OrchardCore.Tests/Data/ContentItemTests.cs:128]: Ah, right

**Comment by hishamco** [comment]
@sarahelsaig waiting for your approval if everything looks good for you

**Review by sarahelsaig — APPROVED** [review]
LGTM

**Comment by hishamco** [comment]
@MikeAlhayek is there anything to add or shall we merge this

**Comment by MikeAlhayek** [comment]
Merge it

**Pull request merged**
